File: composeApp/src/iosMain/kotlin/no/nordicsemi/nrf/matter/matter/CommandExecutor.kt
package no.nordicsemi.nrf.matter.matter

import kotlinx.cinterop.ExperimentalForeignApi
import platform.Foundation.NSNumber
import platform.Matter.MTRBaseDevice
import platform.Matter.MTRContextTagKey
import platform.Matter.MTRDataKey
import platform.Matter.MTRStructureValueType
import platform.Matter.MTRTypeKey
import platform.Matter.MTRValueKey

/** One TLV-tagged field of a Matter command invocation, keyed by its spec-defined context tag. */
internal data class CommandField(val contextTag: Int, val type: String, val value: Any)

/**
 * Mirrors `iosApp/iosApp/kotlin/local/CommandExecutor.swift`. Generalized to support
 * multi-field (and zero-field) command invocations, needed for commands like
 * "Move to Level with On/Off" that the original Swift only used with a single field.
 */
@OptIn(ExperimentalForeignApi::class)
internal class CommandExecutor(deviceId: NSNumber) {

    private val baseDevice: MTRBaseDevice = MTRBaseDevice.deviceWithNodeID(
        nodeID = deviceId,
        controller = LocalControllerProviderImpl.getController("CommandExecutor"),
    )

    suspend fun executeCommand(
        endpoint: NSNumber,
        cluster: NSNumber,
        command: NSNumber,
        type: String,
        value: Any,
    ) = executeCommand(endpoint, cluster, command, listOf(CommandField(0, type, value)))

    suspend fun executeCommand(
        endpoint: NSNumber,
        cluster: NSNumber,
        command: NSNumber,
        fields: List<CommandField> = emptyList(),
    ) {
        val fieldValues = mapOf(
            MTRTypeKey to MTRStructureValueType,
            MTRValueKey to fields.map { field ->
                mapOf(
                    MTRContextTagKey to NSNumber(int = field.contextTag),
                    MTRDataKey to mapOf(
                        MTRTypeKey to field.type,
                        MTRValueKey to field.value,
                    ),
                )
            },
        )

        mtrCall { completion ->
            baseDevice.invokeCommandWithEndpointID(
                endpointID = endpoint,
                clusterID = cluster,
                commandID = command,
                commandFields = fieldValues,
                timedInvokeTimeout = null,
                queue = defaultMatterQueue(),
                completion = completion,
            )
        }
    }
}

File: iosApp/iosDeps/local/CommandExecutor.swift
//
//  CommandExecutor.swift
//  iosApp
//
//  Created by Sylwester Zielinski on 23/04/2026.
//

import Matter

/// Invokes cluster commands on a Matter device over the local (on-device) controller.
class CommandExecutor {

    private let baseDevice: MTRBaseDevice

    /// Creates an executor for the device with the given node ID.
    ///
    /// - Parameter deviceId: The Matter node ID of the target device.
    /// - Throws: An error if the local controller cannot be obtained.
    init(deviceId: NSNumber) throws {
        let controller = try LocalControllerProvider(logTag: "AttributeWriter").getController()
        baseDevice = MTRBaseDevice(nodeID: deviceId, controller: controller)
    }

    /// Invokes a cluster command with a single structured field value.
    ///
    /// - Parameters:
    ///   - endpoint: The endpoint ID hosting the cluster.
    ///   - cluster: The cluster ID the command belongs to.
    ///   - command: The command ID to invoke.
    ///   - type: The MTR value type of the command field (e.g. `MTRBooleanValueType`).
    ///   - value: The value to send for the command field.
    /// - Throws: An error if the command invocation fails.
    func executeCommand(endpoint: NSNumber, cluster: NSNumber, command: NSNumber, type: String, value: Any) async throws {
        SharedLogger.debug("Executing command: \(command)")

        let fields: NSDictionary = [
            MTRTypeKey: MTRStructureValueType,
            MTRValueKey: [
                [
                    MTRContextTagKey: 0,
                    MTRDataKey: [
                        MTRTypeKey: type,
                        MTRValueKey: value
                    ]
                ]
            ]
        ]
        
        try await baseDevice.invokeCommand(withEndpointID: endpoint, clusterID: cluster, commandID: command, commandFields: fields, timedInvokeTimeout: nil, queue: DispatchQueue.global())
        
        SharedLogger.debug("Command executed successfully.")
    }
}
